### PR TITLE
Add e2e tests for Federated Ingress

### PR DESCRIPTION
**Pull request by mfanjie**
This is e2e code for federation ingress controller.
Based util functions, add federation ingress e2e cases(reuse current k8s ones) and add logic to validate the result.

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/29773)

### CODE REVIEW

**Comment by mfanjie** [comment]
@quinton-hoole

**Comment by quinton-hoole** [comment]
Thanks @mfanjie .  Please factor out all of the code cut 'n pasted from the federated services test into common libraries to avoid code duplication.

Please also incorporate the existing ingress tests:

https://github.com/kubernetes/kubernetes/blob/master/test/e2e/ingress.go

Similarly, without cutting and pasting code.

One option would be to have both federated and non-federated ingress tests as one set of tests, with the federation-specific parts masked out by `framework.SkipUnlessFederated()`.  We definitely want to have all of the non-federated tests run against the federation to make sure that the federated API confirms with the non-federated one at all times.

**Comment by mfanjie** [comment]
@quinton-hoole I am revising e2e test code, the current goal is the run the conformanceTests defined in ingress e2e in federation-ingress e2e. While I doubt if we can combine ingress e2e and federation e2e together, as even the cases are same, the methods of checking result are different.
I will extract those common methods to util file, but my current approach is still using different go file for federation ingress e2e. 
The discussion is still open, I can do it later when I find it is doable and appropriate.

**Comment by mfanjie** [comment]
Major changes:
1. extract common code from e2e/ingress.go to ingress_util.go, we plan to run same e2e test cases on kubernetes and on federation, and reuse the function of handling Ingress Objects. 
This include some changes to talk to apiserver, as ingress e2e hold the `unversioned.Client`, I tend to change this client to `RestClient` so that it can talks to both kubernetes apiserver and federation apisever.
2. extract common code from e2e/federated-service.go to federation_util.go, plan to reuse the function of handling cluster setup and teardown.
3. based on bullet 1 and 2, add federation ingress e2e cases(reuse current k8s ones) and add logic to validate the result.

**Comment by mfanjie** [comment]
more code changes are made, the first step is to make sure the refactor works as expected, no side effect of existing e2e.

**Comment by mfanjie** [comment]
@quinton-hoole As most of the changes of this PR is refactoring the existing e2e code and extract the common function to utils, I would suggest to merge this portion first after we verified and confirmed no side defect introduced.
And  then I can focus on the federation ingress cases, which is blocked by ingress resources support on federation apiserver, and our TBD ingress controller.
Are you fine with it?

**Comment by quinton-hoole** [comment]
Thanks @mfanjie .  Catching up from vacation so please excuse brevity. 
I don't think that Federation Ingress cases are blocked.  They can be developed now, and will simply fail until all of the required functionality that you mention is in place.
And yes, I'm fine with temporarily having two sets of tests, one refactored and the other not, and deleting the older, non-refactored version as soon as we've confirmed that there are no regressions with that.

PS: I've not reviewed this PR yet, so I hope that the above comments make sense.

**Comment by mfanjie** [comment]
@quinton-hoole refactor existing code is the first step, and I think it has been done in this PR.
What I proposed is I want to sperated this PR to two, 
1. one PR is refactoring for existing e2e code(federation-gress.go should be excluded as it is newly added), after your code review and after I confirm no regression was reduced, we can merge it ASAP, this reduce the potential conflicts with others, focus would be existing service e2e, ingress e2e, federation e2e 
2. then we can focusing on federation ingress e2e, which is mainly federation-ingress.go, only focus on federation ingress e2e, thing becomes easier..

File: test/e2e/federated-service.go [hunk 1]
```diff
@@ -41,19 +37,13 @@ import (
 )
 
 const (
-	UserAgentName = "federation-e2e-service-controller"
-	// TODO(madhusudancs): Using the same values as defined in the federated
-	// service controller. Replace it with the values from the e2e framework.
-	KubeAPIQPS   = 20.0
-	KubeAPIBurst = 30
+	SCUserAgentName = "federation-e2e-service-controller"
```
Comment (quinton-hoole) [inline test/e2e/federated-service.go]: Do we really need two different user agent names?

File: test/e2e/federation-ingress.go [hunk 2]
```diff
@@ -20,285 +20,128 @@ import (
...
 
-			//TODO for ingress, the controller does only rely on pod status only, need add more objects creation and check
 			BeforeEach(func() {
 				framework.SkipUnlessFederated(f.Client)
-				createBackendPodsOrFail(clusters, f.Namespace.Name, FederatedIngressPodName)
-				ingress = createIngressOrFail(f.FederationClientset_1_3, f.Namespace.Name)
-				waitForIngressShardsOrFail(f.Namespace.Name, ingress, clusters)
+				framework.SkipUnlessProviderIs("gce", "gke")
+				By("Initializing gce controller")
+				gceController = &GCEIngressController{ns: ns, Project: framework.TestContext.CloudConfig.ProjectID, c: jig.client}
```
Comment (quinton-hoole) [inline test/e2e/federation-ingress.go]: I don't think that we want the GCE ingress controller to be created in these e2e tests.  The controller is already automatically created in GKE and GCE clusters.  At worst we should edit the scripts that bring up the GCE/GKE test clusters to include the GCE ingress controller.  I would ideally like these tests to run correctly against any cluster type/ingress controller type (with perhaps additional checks per ingress controller type where necessary).

File: test/e2e/federation-ingress.go [hunk 3]
```diff
@@ -20,285 +20,128 @@ import (
...
 				framework.SkipUnlessFederated(f.Client)
-				deleteBackendPodsOrFail(clusters, f.Namespace.Name)
-
-				if ingress != nil {
-					deleteIngressOrFail(f.FederationClientset_1_3, f.Namespace.Name, ingress.Name)
-					ingress = nil
-				} else {
-					By("No ingress to delete.  Ingress is nil")
+				if CurrentGinkgoTestDescription().Failed {
+					describeIng(ns)
```
Comment (quinton-hoole) [inline test/e2e/federation-ingress.go]: This is confusing.  Presumably there are some side-effects of describeIng that you want to invoke here, but the code does not make that clear.  Please restructure the code accordingly (e.g. with appropriate return values for describeIng() )

File: test/e2e/federation-ingress.go [hunk 4]
```diff
@@ -311,15 +154,15 @@ var _ = framework.KubeDescribe("[Feature:Federation]", func() {
...
-	clusterIngress.Spec.ExternalIPs = federationIngress.Spec.ExternalIPs
-	clusterIngress.Spec.DeprecatedPublicIPs = federationIngress.Spec.DeprecatedPublicIPs
-	clusterIngress.Spec.LoadBalancerIP = federationIngress.Spec.LoadBalancerIP
-	clusterIngress.Spec.LoadBalancerSourceRanges = federationIngress.Spec.LoadBalancerSourceRanges
-	// N.B. We cannot iterate over the port objects directly, as their values
-	// only get copied and our updates will get lost.
-	for i := range clusterIngress.Spec.Ports {
-		clusterIngress.Spec.Ports[i].NodePort = federationIngress.Spec.Ports[i].NodePort
-	}
+	//clusterIngress.Spec.ExternalIPs = federationIngress.Spec.ExternalIPs
```
Comment (quinton-hoole) [inline test/e2e/federation-ingress.go]: Remove dead code.  Also I don't think that this equivalence operator currently works correctly given all of the commented out code?

File: test/e2e/ingress_utils.go [hunk 5]
```diff
@@ -151,14 +281,413 @@ func createSecret(kubeClient *client.Client, ing *extensions.Ingress) (host stri
...
+		}
+		framework.Failf("Failed to cleanup GCE L7 resources.")
+	}
+}
+
+// Cleanup cleans up cloud resources.
+// If del is false, it simply reports existing resources without deleting them.
+// It always deletes resources created through it's methods, like staticIP, even
+// if del is false.
+func (cont *GCEIngressController) Cleanup(del bool) error {
```
Comment (quinton-hoole) [inline test/e2e/ingress_utils.go]: This is a very long function.  Would it make sense to break it up into smaller sub-functions?

File: test/e2e/federation-util.go [hunk 6]
```diff
@@ -35,9 +35,9 @@ import (
 )
 
 var (
-	KubeAPIQPS   = 20.0
-	KubeAPIBurst = 30
-	DefaultFederationName = "federation"
+	KubeAPIQPS            float32 = 20.0
```
Comment (quinton-hoole) [inline test/e2e/federation-util.go]: Why this change?   Just curious.

**Comment by quinton-hoole** [comment]
@mfanjie Thanks! Most of the tests you created in the first commit were approximately what is required, but you seem to have removed them in subsequent commits?

1. "Ingress creation should succeed", 
2. "Ingress creation should create matching ingresss in underlying clusters"
3. "... should be able to create a federated ingress- non-local federated ingress"

... etc 

Why did you remove these?  Those cases don't seem to be covered?

**Comment by quinton-hoole** [comment]
@mfanjie Let me know when you've addressed my comments and are ready for the next review.  Yes, I agree, splitting this into two PR's (one for the refactor, and the other to add the ingress e2e tests) makes a lot of sense.

File: test/e2e/federated-service.go (hunk 1 above)
Comment (mfanjie) [inline test/e2e/federated-service.go]: updated federation user agent name to `federation-e2e`, and this can be reused by all federation e2e code

File: test/e2e/federation-ingress.go (hunk 2 above)
Comment (mfanjie) [inline test/e2e/federation-ingress.go]: So from federation level, we assume the kubernetes ingress working properly, and just validate the underlying ingress object status? Is that enough?

File: test/e2e/federation-ingress.go (hunk 3 above)
Comment (mfanjie) [inline test/e2e/federation-ingress.go]: This is a function defined in https://github.com/kubernetes/kubernetes/blob/master/test/e2e/ingress.go#L190, which was extracted to ingress-utils.go, the only purpose is to dump ingress object info in case e2e test fail.

File: test/e2e/federation-ingress.go (hunk 4 above)
Comment (mfanjie) [inline test/e2e/federation-ingress.go]: done, reflect.DeepEqual() should be able to handle the entire spec equivalence.

File: test/e2e/ingress_utils.go (hunk 5 above)
Comment (mfanjie) [inline test/e2e/ingress_utils.go]: done

**Comment by mfanjie** [comment]
@quinton-hoole the first commit is based on federated-service.go, so you saw the similar tests with federated-service.go, but later on, as the previous discussion, we want to combine existing ingress tests into federation ingress, so I did modification based on ingress.go.

For example, `It("should succeed", func() {` should have cover the ingress creation and validation logic. `t.execute()` handles the ingress creation, and `waitForIngressShardsOrFail(f.Namespace.Name, jig.ing, clusters)` valid the ingress creation.

I think this should adapt what kind of validation we need to do on federation ingress e2e, if we use it to valid the ingress shards, then we can revert the case like in 1st commit. If we want to create real backend service, pod, and validate the traffic, we should go current code path.

**Comment by mfanjie** [comment]
@quinton-hoole cloud you review the cases again? thanks.

**Comment by quinton-hoole** [comment]
@mfanjie I'm a bit confused.  You were going to split this into two PR's?  One for the refactor of the federated-service tests, and another for the addition of the federated-ingress tests?  This still looks like one single PR which combines both of the above.  Could you please separate into 2 separate PR's as discussed?  This one is very difficult to review.  I'd rather merge the refactor PR separately, so that we can confirm no regressions.

**Comment by mfanjie** [comment]
@quinton-hoole please review #30671 for extracting common function PR.
for federation ingress e2e PR, please only focus on federation-ingress.go, the will will be rebased when #30671 is merged

**Comment by quinton-hoole** [comment]
Reviewed 1 of 6 files at r3, 3 of 5 files at r4.
Review status: 4 of 6 files reviewed at latest revision, 10 unresolved discussions.

---

*[test/e2e/federation-ingress.go, line 21 \[r4\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/29773#-KPOE8TjvW_FLiGY667i:-KPOE8TjvW_FLiGY667j:b-9xpfam) ([raw file](https://github.com/kubernetes/kubernetes/blob/775b85f56bded58d81b2f9466159a42632e0a20a/test/e2e/federation-ingress.go#L21)):*
> ```Go
> import (
> 	"fmt"
> 	"os"
> ```

Please rename this source file to "federated-ingress.go" to be consistent with others.

---

*[test/e2e/federation-ingress.go, line 42 \[r4\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/29773#-KPOEJii6IhCnmDjxHkA:-KPOEJijfDeuNBU1TPQ9:b-2jshp7) ([raw file](https://github.com/kubernetes/kubernetes/blob/775b85f56bded58d81b2f9466159a42632e0a20a/test/e2e/federation-ingress.go#L42)):*
> ```Go
> const (
> 	FederatedIngressTimeout        = 60 * time.Second
> 	FederationIngressName          = "federation-ingress"
> ```

Change "Federation" to "Federated" throughout, except where referring specifically to the federation of clusters.

---

*[test/e2e/federation-ingress.go, line 96 \[r4\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/29773#-KPOFIyRKEUIySleylhD:-KPOFIyS9hUql3ZSuOFQ:b-lrzj88) ([raw file](https://github.com/kubernetes/kubernetes/blob/775b85f56bded58d81b2f9466159a42632e0a20a/test/e2e/federation-ingress.go#L96)):*
> ```Go
> 			})
> 
> 			It("should create matching ingresses in underlying clusters", func() {
> ```

Please add tests for updates, and deletes.  In the case of delete we've decided to delete the federated ingress, but leave the ingresses in the underlying clusters intact, for now.  In future we will add a "cascading delete" option, which will also delete the underlying ingresses. 

---

*[test/e2e/framework/util.go, line 4148 \[r4\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/29773#-KPODuOvc6h8t8SyOM7o:-KPODuOvc6h8t8SyOM7p:b-799ygf) ([raw file](https://github.com/kubernetes/kubernetes/blob/775b85f56bded58d81b2f9466159a42632e0a20a/test/e2e/framework/util.go#L4148)):*
> ```Go
> 	var ing *extensions.Ingress
> 	var ok bool
> 	o, err := client.Get().Namespace(ns).Resource("ingress").Name(name).Do().Get()
> ```

The original here is better.  Please revert.

---

*[test/e2e/framework/util.go, line 4152 \[r4\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/29773#-KPOE-rpWUlr2iMD0_m2:-KPOE-rpWUlr2iMD0_m3:byui8j2) ([raw file](https://github.com/kubernetes/kubernetes/blob/775b85f56bded58d81b2f9466159a42632e0a20a/test/e2e/framework/util.go#L4152)):*
> ```Go
> 		return nil, err
> 	}
> 	if ing, ok = o.(*extensions.Ingress); !ok {
> ```

As above.  Remove this.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/29773)*

**Comment by quinton-hoole** [comment]
Reviewed 1 of 5 files at r4.
Review status: 5 of 6 files reviewed at latest revision, 8 unresolved discussions.

---

*[test/e2e/federation-ingress.go, line 94 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/29773#-KOg_w2MqzL8keyFD8PZ-r2-94:-KPOL2oaY0Q_n_OZ8WKC:b-do8ko5) ([raw file](https://github.com/kubernetes/kubernetes/blob/70850fecade99cb524ba59751ca8d315603c9bfb/test/e2e/federation-ingress.go#L94)):*
<details><summary><i>Previously, mfanjie (Jesse Meng) wrote…</i></summary>
> So from federation level, we assume the kubernetes ingress working properly, and just validate the underlying ingress object status? Is that enough? the gceController here is used to obtain static ip and then validate the traffic of the ingress. 
> 
> If we can trust kubernetes ingress controller behaving correctly, and we can drop ingress traffic validation here and then we can remove the gceController here.
</details>
No, sorry for the confusion.  We do need to validate that the DNS name and IP address of the federated ingress are resolvable and connectable.  But the values of those (DNS and IP) can be read from the federated ingress object (they will be populated by the federated ingress controller).  You do not need to access the underlying cloud API here in the tests for that.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/29773)*

**Comment by quinton-hoole** [comment]
Thanks @mfanjie , review complete.  Once the issues are addressed, LGTM.

**Comment by quinton-hoole** [comment]
@mfanjie Do you think you will be able to have this complete by the end of this week?

**Comment by mfanjie** [comment]
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[test/e2e/federation-ingress.go, line 21 \[r4\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/29773#-KPOE8TjvW_FLiGY667i:-KPWXTkPR23D0WJNkWuo:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/775b85f56bded58d81b2f9466159a42632e0a20a/test/e2e/federation-ingress.go#L21)):*
<details><summary><i>Previously, quinton-hoole (Quinton Hoole) wrote…</i></summary>
> Please rename this source file to "federated-ingress.go" to be consistent with others.
</details>
Done.

---

*[test/e2e/federation-ingress.go, line 42 \[r4\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/29773#-KPOEJii6IhCnmDjxHkA:-KPWXW0S51ExOpDr2Hmg:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/775b85f56bded58d81b2f9466159a42632e0a20a/test/e2e/federation-ingress.go#L42)):*
<details><summary><i>Previously, quinton-hoole (Quinton Hoole) wrote…</i></summary>
> Change "Federation" to "Federated" throughout, except where referring specifically to the federation of clusters.
</details>
Done.

---

*[test/e2e/federation-ingress.go, line 96 \[r4\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/29773#-KPOFIyRKEUIySleylhD:-KPWXvU8PQk-Sbzs0E1s:bek87e7) ([raw file](https://github.com/kubernetes/kubernetes/blob/775b85f56bded58d81b2f9466159a42632e0a20a/test/e2e/federation-ingress.go#L96)):*
<details><summary><i>Previously, quinton-hoole (Quinton Hoole) wrote…</i></summary>
> Please add tests for updates, and deletes.  In the case of delete we've decided to delete the federated ingress, but leave the ingresses in the underlying clusters intact, for now.  In future we will add a "cascading delete" option, which will also delete the underlying ingresses. 
</details>
the first It is create and delete, the second It is create and validate underlying ingress shard.
I will add another It regarding to update, but it requires the ingress being created in advanced.
So now we have two It, create and delete, create and update.
We can discuss this further later if you have other thought.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/29773)*

**Comment by mfanjie** [comment]
Review status: all files reviewed at latest revision, 9 unresolved discussions.

---

*[test/e2e/federation-ingress.go, line 94 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/29773#-KOg_w2MqzL8keyFD8PZ-r2-94:-KPWZR2j9NQBSXFqovZM:bws1yym) ([raw file](https://github.com/kubernetes/kubernetes/blob/70850fecade99cb524ba59751ca8d315603c9bfb/test/e2e/federation-ingress.go#L94)):*
<details><summary><i>Previously, quinton-hoole (Quinton Hoole) wrote…</i></summary>
> No, sorry for the confusion.  We do need to validate that the DNS name and IP address of the federated ingress are resolvable and connectable.  But the values of those (DNS and IP) can be read from the federated ingress object (they will be populated by the federated ingress controller).  You do not need to access the underlying cloud API here in the tests for that.
</details>
I read your PR, but did not see code of setting dns to ingress object, so I just discoverService as is, which is using nslookup. it can be changed later.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/29773)*

**Comment by mfanjie** [comment]
@quinton-hoole as we only need to take action on federated-ingress.go on this PR, I tried to modify the cases based on your comments, and ran e2e test locally, but without your PR being merged, the test failed.
Please review if I grasp your points in the new committed code.

We will have an team outing this Sunday and next Monday, so I am not sure if we can finish this PR before Sunday. 
If you want to catch some deadline during my absence, please feel free to get the code and submit an new PR if you like.

**Comment by mfanjie** [comment]
As discussed with @quinton-hoole , there are three It() in the federated-ingress.go
1. `should be created and deleted successfully`
This one is created by @nikhiljindal and the purpose is to test ingress object creation and deletion federation apiserver
2. `should create and update matching ingresses in underlying clusters`
This it is to test the ingress shard in underlying clusters have been synced correctly with federation ingress, test the spec sync of federation ingress controller
3. `Ingress connectivity and DNS`
This one is to test the connectivity of the federation ingress, and corresponding dns records in underlying k8s cluster and global dns.
As discussed, this It is marked as pending, so it won't fail the e2e.

**Comment by mfanjie** [comment]
ping @quinton-hoole

**Comment by quinton-hoole** [comment]
Review status: 0 of 4 files reviewed at latest revision, 16 unresolved discussions.

---

*[test/e2e/federated-ingress.go, line 48 \[r5\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/29773#-KQ6X6rOQbCjpNwt1ZOO:-KQ6X6rOQbCjpNwt1ZOP:b-8k61d8) ([raw file](https://github.com/kubernetes/kubernetes/blob/9982edb39589182c081b6e09adf2abac5dca091a/test/e2e/federated-ingress.go#L48)):*
> ```Go
> )
> 
> var _ = framework.KubeDescribe("Federation ingresses [Feature:Federation]", func() {
> ```

nit: Federated ingresses, not Federation ingresses

---

*[test/e2e/federated-ingress.go, line 55 \[r5\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/29773#-KQ6Xe-AePFxYxAPhcJ5:-KQ6Xe-AePFxYxAPhcJ6:b-ca5kt4) ([raw file](https://github.com/kubernetes/kubernetes/blob/9982edb39589182c081b6e09adf2abac5dca091a/test/e2e/federated-ingress.go#L55)):*
> ```Go
> 	Describe("Ingress objects", func() {
> 		AfterEach(func() {
> 			framework.SkipUnlessFederated(f.Client)
> ```

This belongs in BeforeEach(), not AfterEach(), right?  You want to skip the whole test if not running against a federation.

---

*[test/e2e/federated-ingress.go, line 68 \[r5\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/29773#-KQ6Y6tj_vG6DALD6qQW:-KQ6Y6tj_vG6DALD6qQX:b7iy1jn) ([raw file](https://github.com/kubernetes/kubernetes/blob/9982edb39589182c081b6e09adf2abac5dca091a/test/e2e/federated-ingress.go#L68)):*
> ```Go
> 
> 		It("should be created and deleted successfully", func() {
> 			framework.SkipUnlessFederated(f.Client)
> ```

Aah, I see what you're doing.  This needs to be lifted into BeforeEach(), as all of the tests need to be skipped if not federated.

---

*[test/e2e/federated-ingress.go, line 181 \[r5\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/29773#-KQBH23y3PQdVf0vz7o-:-KQBH23zTOH1L2v1F3Rc:b-ay3ub9) ([raw file](https://github.com/kubernetes/kubernetes/blob/9982edb39589182c081b6e09adf2abac5dca091a/test/e2e/federated-ingress.go#L181)):*
> ```Go
> /*
>    equivalent returns true if the two ingresss are equivalent.  Fields which are expected to differ between
>    federated ingresss and the underlying cluster ingresss (e.g. ClusterIP, LoadBalancerIP etc) are ignored.
> ```

These comments are inconsistent with the code.

---

*[test/e2e/federated-ingress.go, line 225 \[r5\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/29773#-KQBHGG1dh5br7gJG0bX:-KQBHGG1dh5br7gJG0bY:bm5rd24) ([raw file](https://github.com/kubernetes/kubernetes/blob/9982edb39589182c081b6e09adf2abac5dca091a/test/e2e/federated-ingress.go#L225)):*
> ```Go
> 
> /*
>    waitForIngressShardsOrFail waits for the ingress to appear in all clusters
> ```

This comment is cut and pasted from above, but is inconsistent with this code.

---

*[test/e2e/federated-ingress.go, line 235 \[r5\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/29773#-KQBHPVtlzsRh_n-C7PS:-KQBHPVtlzsRh_n-C7PT:bm5rd24) ([raw file](https://github.com/kubernetes/kubernetes/blob/9982edb39589182c081b6e09adf2abac5dca091a/test/e2e/federated-ingress.go#L235)):*
> ```Go
> 
> /*
>    waitForIngressOrFail waits until a ingress is either present or absent in the cluster specified by clientset.
> ```

This comment is cut and pasted from above, but is inconsistent with this code.

---

*[test/e2e/federated-ingress.go, line 315 \[r5\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/29773#-KQBHzEVDcP6AonRPBEp:-KQBHzEVDcP6AonRPBEq:b-y5fmb3) ([raw file](https://github.com/kubernetes/kubernetes/blob/9982edb39589182c081b6e09adf2abac5dca091a/test/e2e/federated-ingress.go#L315)):*
> ```Go
> 	}
> 
> 	for i := 0; i < 3; i++ {
> ```

Please lift the number of retries into a named constant, rather than have the magic number 3 in here as a literal.  Also, it's better to use the util.Poll() library function.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/29773)*

**Comment by quinton-hoole** [comment]
Reviewed 6 of 7 files at r5.
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/29773)*

**Comment by quinton-hoole** [comment]
Thanks @mfanjie .  One last set of changes and then LGTM.

**Comment by mfanjie** [comment]
Review status: all files reviewed at latest revision, 13 unresolved discussions.

---

*[test/e2e/federated-ingress.go, line 48 \[r5\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/29773#-KQ6X6rOQbCjpNwt1ZOO:-KQJ585g_6-vFXJqazxH:bs7wumb) ([raw file](https://github.com/kubernetes/kubernetes/blob/9982edb39589182c081b6e09adf2abac5dca091a/test/e2e/federated-ingress.go#L48)):*
<details><summary><i>Previously, quinton-hoole (Quinton Hoole) wrote…</i></summary>
> nit: Federated ingresses, not Federation ingresses
</details>
Done.

---

*[test/e2e/federated-ingress.go, line 55 \[r5\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/29773#-KQ6Xe-AePFxYxAPhcJ5:-KQJ6eaFI4EfulDaaI50:bj8glyw) ([raw file](https://github.com/kubernetes/kubernetes/blob/9982edb39589182c081b6e09adf2abac5dca091a/test/e2e/federated-ingress.go#L55)):*
<details><summary><i>Previously, quinton-hoole (Quinton Hoole) wrote…</i></summary>
> This belongs in BeforeEach(), not AfterEach(), right?  You want to skip the whole test if not running against a federation.
</details>
This test comes from current federation-ingress.go, I will remove this line in AfterEach().

---

*[test/e2e/federated-ingress.go, line 68 \[r5\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/29773#-KQ6Y6tj_vG6DALD6qQW:-KQJ7HyTZdK9-FC_jEAv:bd6s03d) ([raw file](https://github.com/kubernetes/kubernetes/blob/9982edb39589182c081b6e09adf2abac5dca091a/test/e2e/federated-ingress.go#L68)):*
<details><summary><i>Previously, quinton-hoole (Quinton Hoole) wrote…</i></summary>
> Aah, I see what you're doing.  This needs to be lifted into BeforeEach(), as all of the tests need to be skipped if not federated.
</details>
ok, I will keep framework.SkipUnlessFederated(f.Client) only in BeforeEach, but if you look at federation-service.go, it is everywhere, Before, After and It. that file should be addressed later.

---

*[test/e2e/federated-ingress.go, line 181 \[r5\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/29773#-KQBH23y3PQdVf0vz7o-:-KQJ8KVe1sw_KghWZXsF:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/9982edb39589182c081b6e09adf2abac5dca091a/test/e2e/federated-ingress.go#L181)):*
<details><summary><i>Previously, quinton-hoole (Quinton Hoole) wrote…</i></summary>
> These comments are inconsistent with the code.
</details>
Done.

---

*[test/e2e/federated-ingress.go, line 225 \[r5\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/29773#-KQBHGG1dh5br7gJG0bX:-KQJ8V01BKVR6cRAwT0c:b3y4gq7) ([raw file](https://github.com/kubernetes/kubernetes/blob/9982edb39589182c081b6e09adf2abac5dca091a/test/e2e/federated-ingress.go#L225)):*
<details><summary><i>Previously, quinton-hoole (Quinton Hoole) wrote…</i></summary>
> This comment is cut and pasted from above, but is inconsistent with this code.
</details>
done, sharp eyes.

---

*[test/e2e/federated-ingress.go, line 235 \[r5\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/29773#-KQBHPVtlzsRh_n-C7PS:-KQJ8r-UnxKJaMYsN9xB:b-896fix) ([raw file](https://github.com/kubernetes/kubernetes/blob/9982edb39589182c081b6e09adf2abac5dca091a/test/e2e/federated-ingress.go#L235)):*
<details><summary><i>Previously, quinton-hoole (Quinton Hoole) wrote…</i></summary>
> This comment is cut and pasted from above, but is inconsistent with this code.
</details>
Done.

---

*[test/e2e/federated-ingress.go, line 315 \[r5\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/29773#-KQBHzEVDcP6AonRPBEp:-KQJ9dSV-u6msrTYsw3n:bml9jx3) ([raw file](https://github.com/kubernetes/kubernetes/blob/9982edb39589182c081b6e09adf2abac5dca091a/test/e2e/federated-ingress.go#L315)):*
<details><summary><i>Previously, quinton-hoole (Quinton Hoole) wrote…</i></summary>
> Please lift the number of retries into a named constant, rather than have the magic number 3 in here as a literal.  Also, it's better to use the util.Poll() library function.
</details>
changed to MaxRetryOnFederatedApiserver at this moment. util.Poll() change should be in following PR.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/29773)*

**Comment by quinton-hoole** [comment]
LGTM!  Thanks @mfanjie. And thanks for your patience.

**Comment by quinton-hoole** [comment]
ok to test

**Comment by quinton-hoole** [comment]
:lgtm:

---

Reviewed 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 13 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/29773#-:-KQOkE9Pr1NbyecwpsYE:bnfp4nl)*

**Comment by quinton-hoole** [comment]
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[test/e2e/federated-ingress.go, line 315 \[r5\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/29773#-KQBHzEVDcP6AonRPBEp:-KQOl51XE4TSs_7O_cQV:b5x2doz) ([raw file](https://github.com/kubernetes/kubernetes/blob/9982edb39589182c081b6e09adf2abac5dca091a/test/e2e/federated-ingress.go#L315)):*
<details><summary><i>Previously, mfanjie (Jesse Meng) wrote…</i></summary>
> changed to MaxRetryOnFederatedApiserver at this moment. util.Poll() change should be in following PR.
</details>
@mfanje No, you need a constant at the top of the file: MaxRetriesOnFederationAPiServer=3, and then here a for loop `for i:=0; i < 
MaxRetryOnFederatedApiserver etc.

Better still, use one of the existing constants in the e2e/util package, if a suitable one exists.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/29773)*

**Comment by quinton-hoole** [comment]
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[test/e2e/federation-util.go, line 38 \[r3\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/29773#-KOg_w2MqzL8keyFD8Pa-r3-38:-KQOmqjMPgXORB8A8qwP:bb6ywmx) ([raw file](https://github.com/kubernetes/kubernetes/blob/7e55b82cb79a2a9b727d2c305059707b7dbe3152/test/e2e/federation-util.go#L38)):*
<details><summary><i>Previously, quinton-hoole (Quinton Hoole) wrote…</i></summary>
> Why this change?   Just curious.
</details>
LGTM

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/29773)*
